File: python/model/campeonato.py
from typing import Dict, List, Set
from collections import defaultdict
from .time import Time
from .partida import Partida

class Campeonato:
    def __init__(self, ano: int):
        self.ano = ano
        self.times: Dict[str, Time] = {}
        self.partidas: List[Partida] = []
    
    def adicionar_partida(self, partida: Partida):
        """Adiciona uma partida ao campeonato"""
        self.partidas.append(partida)
        
        # Adicionar times se não existirem
        if partida.mandante not in self.times:
            self.times[partida.mandante] = Time(partida.mandante)
        if partida.visitante not in self.times:
            self.times[partida.visitante] = Time(partida.visitante)
    
    def calcular_estatisticas(self):
        """Calcula todas as estatísticas do campeonato"""
        # Reiniciar estatísticas de todos os times
        for time in self.times.values():
            time.reiniciar()
        
        # Agrupar partidas por rodada
        partidas_por_rodada = defaultdict(list)
        for partida in self.partidas:
            partidas_por_rodada[partida.rodada].append(partida)
        
        # Processar cada rodada em ordem
        rodadas = sorted(partidas_por_rodada.keys())
        
        for rodada in rodadas:
            partidas_rodada = partidas_por_rodada[rodada]
            
            # Atualizar estatísticas dos times para esta rodada
            for partida in partidas_rodada:
                time_mandante = self.times[partida.mandante]
                time_visitante = self.times[partida.visitante]
                
                time_mandante.adicionar_resultado_sem_posicao(
                    partida.gols_mandante,
                    partida.gols_visitante,
                    rodada
                )
                
                time_visitante.adicionar_resultado_sem_posicao(
                    partida.gols_visitante,
                    partida.gols_mandante,
                    rodada
                )
            
            # Calcular e atribuir posições na tabela após a rodada
            self._calcular_e_atualizar_posicoes(rodada)
    
    def _calcular_e_atualizar_posicoes(self, rodada: int):
        """Calcula as posições na tabela após uma rodada"""
        tabela = list(self.times.values())
        
        # Ordenar por critérios do brasileirão
        tabela.sort(key=lambda t: (
            -t.pontos,           # Mais pontos primeiro
            -t.vitorias,         # Mais vitórias primeiro
            -t.get_saldo_gols(), # Melhor saldo primeiro
            -t.gols_feitos       # Mais gols feitos primeiro
        ))
        
        # Atualizar posições no histórico de cada time
        for i, time in enumerate(tabela):
            posicao = i + 1
            time.atualizar_posicao_para_rodada(rodada, posicao)
    
    def get_nomes_dos_times(self) -> Set[str]:
        """Retorna nomes de todos os times"""
        return set(self.times.keys())
    
    def get_time(self, nome: str) -> Time:
        """Retorna um time pelo nome"""
        return self.times.get(nome)

File: java/src/main/java/model/Campeonato.java
package model;

import java.util.*;
import java.util.stream.Collectors;

public class Campeonato {
    private int ano;
    private Map<String, Time> times;
    private List<Partida> partidas;
    
    public Campeonato(int ano) {
        this.ano = ano;
        this.times = new HashMap<>();
        this.partidas = new ArrayList<>();
    }
    
    public void adicionarPartida(Partida partida) {
        partidas.add(partida);
        
        // Adicionar times se não existirem
        times.putIfAbsent(partida.getMandante(), new Time(partida.getMandante()));
        times.putIfAbsent(partida.getVisitante(), new Time(partida.getVisitante()));
    }
    
    public void calcularEstatisticas() {
        // Reiniciar estatísticas de todos os times
        for (Time time : times.values()) {
            time.reiniciar();
        }
        
        // Agrupar partidas por rodada
        Map<Integer, List<Partida>> partidasPorRodada = partidas.stream()
            .collect(Collectors.groupingBy(Partida::getRodada));
        
        // Processar cada rodada em ordem
        List<Integer> rodadas = new ArrayList<>(partidasPorRodada.keySet());
        Collections.sort(rodadas);
        
        for (int rodada : rodadas) {
            List<Partida> partidasRodada = partidasPorRodada.get(rodada);
            
            // Atualizar estatísticas dos times para esta rodada
            for (Partida partida : partidasRodada) {
                Time timeMandante = times.get(partida.getMandante());
                Time timeVisitante = times.get(partida.getVisitante());
                
                // CORREÇÃO: Adicionar resultado SEM a posição (será calculada depois)
                timeMandante.adicionarResultadoSemPosicao(
                    partida.getGolsMandante(), 
                    partida.getGolsVisitante(), 
                    rodada
                );
                
                timeVisitante.adicionarResultadoSemPosicao(
                    partida.getGolsVisitante(), 
                    partida.getGolsMandante(), 
                    rodada
                );
            }
            
            // Calcular e atribuir posições na tabela após a rodada
            calcularEAtualizarPosicoes(rodada);
        }
    }
    
    private void calcularEAtualizarPosicoes(int rodada) {
        List<Time> tabela = new ArrayList<>(times.values());
        
        // Ordenar por critérios do brasileirão
        tabela.sort((t1, t2) -> {
            if (t1.getPontos() != t2.getPontos()) {
                return Integer.compare(t2.getPontos(), t1.getPontos());
            }
            if (t1.getVitorias() != t2.getVitorias()) {
                return Integer.compare(t2.getVitorias(), t1.getVitorias());
            }
            if (t1.getSaldoGols() != t2.getSaldoGols()) {
                return Integer.compare(t2.getSaldoGols(), t1.getSaldoGols());
            }
            return Integer.compare(t2.getGolsFeitos(), t1.getGolsFeitos());
        });
        
        // Atualizar posições no histórico de cada time
        for (int i = 0; i < tabela.size(); i++) {
            Time time = tabela.get(i);
            int posicao = i + 1;
            
            // CORREÇÃO: Atualizar a posição para a rodada específica
            time.atualizarPosicaoParaRodada(rodada, posicao);
        }
    }
    
    public Set<String> getNomesDosTimes() {
        return new HashSet<>(times.keySet());
    }
    
    public Time getTime(String nome) {
        return times.get(nome);
    }
    
    // Getters
    public int getAno() { return ano; }
    public List<Partida> getPartidas() { return new ArrayList<>(partidas); }
}